lib/definition/types/keys.rb

```diff
@@ -57,114 +57,93 @@ def ensure_keys_do_not_interfere(other)
-        self.required_definitions = req.map { |key, definition| { key: key, definition: definition } }
-        self.optional_definitions = opt.map { |key, definition| { key: key, definition: definition } }
+        self.required_definitions = req
+        self.optional_definitions = opt
         self.defaults = defaults
         self.ignore_extra_keys = options.fetch(:ignore_extra_keys, false)
       end
 
-      def conform(value)
-        Conformer.new(self, value).conform
-      end
+      def conform(input_value) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+        # input_value is duplicated because we don't want to modify the user object that is passed into this function.
+        # The following logic will iterate over each definition and delete the key associated with the definition from
+        # the input value.
+        # In the end if there are still keys on the object and 'ignore_extra_keys' is false, an error will be raised.
+        value = input_value.dup
+        result_value = {}
+        errors = []
 
-      def keys
-        (required_definitions + optional_definitions).map { |hash| hash[:key] }
-      end
+        return wrong_type_result(value) unless value.is_a?(Hash)
 
-      class Conformer
-        def initialize(definition, value)
-          self.definition = definition
-          self.value = value
-          self.errors = []
-          @conform_result_value = {} # This will be the output value after conforming
-          @not_conformed_value_keys = value.dup # Used to track which keys are left over in the end (unexpected keys)
-        end
+        required_definitions.each do |key, definition|
+          if value.key?(key)
+            result = definition.conform(value.delete(key))
+            result_value[key] = result.value
+            next if result.passed?
 
-        def conform
-          return invalid_input_result unless valid_input_type?
-
-          values = conform_all_keys
-          add_extra_key_errors unless definition.ignore_extra_keys
-
-          ConformResult.new(values, errors: errors)
+            errors.push(key_error(definition, key, result))
+          else
+            errors << missing_key_error(key)
+          end
         end
 
-        private
-
-        attr_accessor :errors
+        optional_definitions.each do |key, definition|
+          if value.key?(key)
+            result = definition.conform(value.delete(key))
+            result_value[key] = result.value
+            next if result.passed?
 
-        def invalid_input_result
-          errors = [ConformError.new(definition,
-                                     "#{definition.name} is not a Hash",
-                                     i18n_key: "keys.not_a_hash")]
-          ConformResult.new(value, errors: errors)
-        end
-
-        def valid_input_type?
-          value.is_a?(Hash)
+            errors.push(key_error(definition, key, result))
```

Comment:
Maybe try `<<` instead of `push`.
IIRC `<<` this fastest way to add a element to an array.